### DISCUSSION TITLE
Any plan to add `async`?

### DISCUSSION THREAD

**Discussion by felixf4xu** [c3lang/c3c#1532] [Ideas]
Hi,

I'm reading the docs and it seems to be a very interesting language, but I can't find anything related to `async`.

It's mentioned here: https://c3-lang.org/faq/compare-languages/#in-zig-but-not-in-c3 but it seems to indicate that it does not have async now.

When 0.6.3 is feature complete and async is not there, so what's the plan for `async`?

**Comment by lerno** [comment]
There are no plans to add async at all, and this is completely deliberate.

**Reply by felixf4xu** [reply]
thanks, it would be nice to add it to https://c3-lang.org/faq/rejected-ideas/

**Comment by felixf4xu** [comment]
I don't know how to write a compiler, but I was wondering, can compilers be configurable? like some lint tools (https://eslint.org/docs/latest/use/configure/)

In that way, it's easier to put some features configurable to the users. For example, if someone does not like `goto`, he can configure to disable it; if someone else likes `async`, he can configure to enable it.

It must be difficult because I have only heard one compiler like https://github.com/seanbaxter/circle/blob/master/new-circle/README.md#feature-directive-philosophy

**Reply by ellipse12** [reply]
I don't think its necessarily extremely difficult, I think it just adds a lot of unneeded complexity. It also makes the grammar somewhat ambiguous which can make debugging more difficult. Also with adding async as part of the language, it can make the resultant binary harder to predict. C3 and C both try to be fairly explicit about what is going on (C3 has more abstraction but it still has more control than something like rust) hence the reservation in adding it. Also it is possible to implement asynchronous code without special syntax for it, its just harder.

**Reply by tyoc213** [reply]
@ellipse12  care to provide a little example of that? or link a ref?

**Reply by ellipse12** [reply]
@tyoc213 Of the manual async implementation?

**Reply by tyoc213** [reply]
yeah if possible @ellipse12 !

"Also it is possible to implement asynchronous code without special syntax for it, its just harder."

**Reply by ellipse12** [reply]
@tyoc213 Well in general an async function kind of just gets transformed into a state machine where each call to await is translated into a separate intermediate state. For example, in pseudocode, this
```
async fn void wait(int delay) {
    print("before sleep");
    await sleep(delay);
    print("after sleep");
}
```
Gets translated (roughly) to something like:
```
enum WaitState {
    __Init,
    __Await1,//if sleep returned something you would make this into a tagged union and store it under this tag
    __Done
}
fn void wait(WaitState* self, int delay) {
    switch (*self) {
         case __Init: {
             print("before sleep");
             sleep(delay);
             *self = __Await1;
         },
         case __Await1: {
             print("after sleep");
             *self = __Done;
         },
         case __Done: //this would cause an error if wait gets called with this state since there is nothing left to do
     }
}
```
Which is then used with an async runtime. All of this is fully possible in pretty much any language (including c) it's just really annoying to do for each function (also I wrote this on mobile so if something looks weird that's probably why).

**Reply by ellipse12** [reply]
Also the function's parameters would also probably get stored along with the state in a struct of some kind instead of being passed in. And there's a lot more that you would actually need to do (such as returning a future and handling the future returned from sleep) but this is essentially what is happening.
